usr/src/uts/common/fs/zfs/zio.c

```diff
@@ -2860,25 +2860,24 @@ zio_dva_unallocate(zio_t *zio, zio_gang_node_t *gn, blkptr_t *bp)
  */
 int
 zio_alloc_zil(spa_t *spa, uint64_t txg, blkptr_t *new_bp, blkptr_t *old_bp,
-    uint64_t size, boolean_t use_slog)
+    uint64_t size, boolean_t *slog)
 {
 	int error = 1;
 	zio_alloc_list_t io_alloc_list;
 
 	ASSERT(txg > spa_syncing_txg(spa));
 
 	metaslab_trace_init(&io_alloc_list);
-
-	if (use_slog) {
-		error = metaslab_alloc(spa, spa_log_class(spa), size,
```

Comment:
I am not sure, but maybe it would be better to use the input value of `slog` to decide if the `spa_log_class` allocation attempt should be skipped.  I suggest this only as a micro-optimization to avoid calling `metaslab_alloc` and `spa_config_enter` / `spa_config_exit` in it when there is no slog device.